### PR TITLE
[c++] Cast vertex list to WKB

### PR DESCRIPTION
**Pull request by XanthosXanthopoulos**
When writing data, provide an option to generate the WKB buffer as well as spatial index columns instead of requesting user to provide them. To differentiate against other geometry types, the vertex list Arrow array needs to contain a metadata entry `geometry_type` (only `polygon_ring` supported for now).

### CODE REVIEW

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 1]
```diff
@@ -119,4 +123,234 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+
+//===================================================================
+//= private non-static
+//===================================================================
+
+std::vector<ArrowTable> SOMAGeometryDataFrame::_cast_polygon_vertex_list_to_wkb(
+    ArrowArray* array) {
+    // Initialize a vector to hold all the Arrow tables containing the
+    // transformed geometry data
+    ArrowError error;
```
Comment (jp-dark) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:184]: Move this to right before it's used. (Line 274).

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 2]
```diff
@@ -119,4 +123,234 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+        std::make_unique<ArrowSchema>(ArrowSchema{})));
+
+    NANOARROW_THROW_NOT_OK(ArrowArrayInitFromType(
+        tables.front().first.get(), ArrowType::NANOARROW_TYPE_LARGE_BINARY));
+    NANOARROW_THROW_NOT_OK(ArrowSchemaInitFromType(
+        tables.front().second.get(), ArrowType::NANOARROW_TYPE_LARGE_BINARY));
+    NANOARROW_THROW_NOT_OK(
+        ArrowSchemaSetName(tables.front().second.get(), "soma_geometry"));
+
+    for (auto axis : spatial_axes) {
```
Comment (jp-dark) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:198]:
```suggestion
    for (const auto& axis : spatial_axes) {
```

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 3]
```diff
@@ -119,4 +123,234 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+        ArrowArrayReserve(tables.front().first.get(), wkb_buffer_size));
+    NANOARROW_THROW_NOT_OK(
+        ArrowArrayStartAppending(tables.front().first.get()));
+    for (size_t i = 1; i < tables.size(); ++i) {
+        NANOARROW_THROW_NOT_OK(
+            ArrowArrayReserve(tables[i].first.get(), array->length));
+        NANOARROW_THROW_NOT_OK(ArrowArrayStartAppending(tables[i].first.get()));
+    }
+
+    for (auto& geometry : geometries) {
```
Comment (jp-dark) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:256]:
```suggestion
    for (const auto& geometry : geometries) {
```

**Review by jp-dark — COMMENTED** [review]
I have a couple small recommended changes while I'm working to understand the nanoarrow API.

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 4]
```diff
@@ -119,4 +123,234 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+        geometry::BinaryBuffer wkb = geometry::to_wkb(geometry);
+        geometry::Envelope envelope = geometry::envelope(geometry);
+
+        ArrowBufferView wkb_view;
+        wkb_view.data.data = wkb.data();
+        wkb_view.size_bytes = (int64_t)wkb.size();
```
Comment (jp-dark) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:262]:
```suggestion
        wkb_view.size_bytes = static_cast<int64_t>(wkb.size());
```

**Review by jp-dark — APPROVED** [review]

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 5]
```diff
@@ -119,4 +123,233 @@ uint64_t SOMAGeometryDataFrame::count() {
...
 }
 
+void SOMAGeometryDataFrame::set_array_data(
+    std::unique_ptr<ArrowSchema> arrow_schema,
+    std::unique_ptr<ArrowArray> arrow_array) {
+    std::vector<std::string> spatial_axes = this->spatial_column_names();
+
+    for (auto i = 0; i < arrow_schema->n_children; ++i) {
+        /**
+         * If `soma_geometry` conforms to specific formats automatically convert
```
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:133]:
```suggestion
         * If `soma_geometry` conforms to specific formats, automatically convert
```

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 6]
```diff
@@ -119,4 +123,233 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+         * to WKB and create additional index columns for spatial axes.
+         *
+         * If the `soma_geometry` array is a WKB binary users are expected to
```
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:136]:
```suggestion
         * If the `soma_geometry` array is a WKB binary, users are expected to
```

File: libtiledbsoma/src/soma/soma_geometry_dataframe.cc [hunk 7]
```diff
@@ -119,4 +123,233 @@ uint64_t SOMAGeometryDataFrame::count() {
...
+                NANOARROW_THROW_NOT_OK(ArrowMetadataGetValue(
+                    arrow_schema->children[i]->metadata,
+                    ArrowCharView("geometry_type"),
+                    &out));
+
+                type_metadata = std::string_view(out.data, out.size_bytes);
+            }
+
+            ArrowTable casted_data;
+            if (type_metadata == "polygon_ring") {
```
Comment (johnkerl) [inline libtiledbsoma/src/soma/soma_geometry_dataframe.cc:157]: Please make this a named constant in a header file

**Pull request merged**